### PR TITLE
CA-1654 New table for tracking changes to Billing Accounts on Billing Projects

### PR DESCRIPTION
**Pull request by gpolumbo-broad**
https://broadworkbench.atlassian.net/browse/CA-1654

Created a new table, `BILLING_ACCOUNT_CHANGES`.  Whenever a call is made to update the Billing Account on a Billing Project, we make sure that a record is created in that new table.  

Since this new table requires a "who done it" field, we wanted to isolate which calls to update a Billing Project record require a `userSubjectId` to be just calls to update the Billing Account.  This lead us to do some refactoring to remove the generic `updateBillingProjects` query and instead we have specific methods for updating fields on the Billing Project table. 

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20220407_billing_account_changes_table.xml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                    type="VARCHAR(254)"
+                    remarks="Name of the billing project being updated.">
+                <constraints nullable="false"/>
+            </column>
+            <column name="USER_ID"
+                    type="VARCHAR(254)"
+                    remarks="User ID of the user making the change.">
+                <constraints nullable="false"/>
+            </column>
+            <column name="ORIGINAL_BILLING_ACCOUNT"
```
Comment (ehigham) [inline core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20220407_billing_account_changes_table.xml]: Thinking about it, this isn't the original billing account - it's the previous value. Dunno if that'll cause confusion in the future. Perhaps we could change this to `PREVIOUS_BILLING_ACCOUNT` and propagate this change around?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala [hunk 2]
```diff
@@ -93,48 +94,30 @@ trait CreatingBillingProjectMonitor extends LazyLogging with FutureSupport {
...
-        }
-
-        nextStepFuture.recover {
-          case t: Throwable =>
-            logger.error(s"failure processing new project ${project.projectName.value}", t)
-            project.copy(status = CreationStatuses.Error, message = Option(t.getMessage))
-        }
+    projects.toList.traverse_ { project =>
+      // figure out if the project operation is done yet and set the project status accordingly
+      operationsByProject.get(project.projectName) match {
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala:104]: Suggestion in https://github.com/broadinstitute/rawls/pull/1693

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceBillingAccountMonitor.scala [hunk 3]
```diff
@@ -9,10 +9,9 @@ import com.google.api.services.cloudbilling.model.ProjectBillingInfo
 import com.typesafe.scalalogging.LazyLogging
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceBillingAccountMonitor.scala:9]: Just seems imports are changed - perhaps revert the file if they're not needed?

**Review by ehigham — CHANGES_REQUESTED** [review]
revert BillingApiServiceV2Spec too please!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala [hunk 4]
```diff
@@ -93,48 +94,30 @@ trait CreatingBillingProjectMonitor extends LazyLogging with FutureSupport {
...
+          // project operation finished successfully
+          onSuccessfulProjectCreate(project)
+
+        case Some(Seq(RawlsBillingProjectOperationRecord(_, _, _, true, Some(error), _))) =>
+          // project operation finished with an error
+          onFailedProjectCreate(project, error)
+
+        case Some(tooManyOperations) if tooManyOperations.size > 1 =>
+          // this should be impossible due to database constraints (duplicate primary keys)
+          onFailedProjectCreate(project, s"Only expected one Operation, found $tooManyOperations")
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala:115]: My preference is to not make too many logic changes here.  Yes, we need to make changes inside the `onSuccessfulProjectCreate` and `onFailedProjectCreate` methods, but I'd rather try to keep risk down and not muck around with the format/style of the pattern matching here, especially if this stuff isn't tested all that well.  And it's v1 Billing Project which is deprecated

**Comment by gpolumbo-broad** [comment]
LGTM 👍 I am OK to merge but cannot review because I opened the PR

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala (hunk 4 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala:115]: I've inlined the calls here - I think this is a safe simplification.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala (hunk 2 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/CreatingBillingProjectMonitor.scala:104]: Will change base when this PR gets merged in and we can submit by itself

**Review by ehigham — APPROVED** [review]
Huzzah!

**Comment by gpolumbo-broad** [comment]
@andy7i - Since Edmund and I have both reviewed and thumbed, we are going to merge so that we can move onto the next issue that builds off of this, but please still review it if you want, we can always adjust things

**Pull request merged**